Update Sampler interface base on OTEP 0006-sampling (#296)
open-telemetry/opentelemetry-specification: specification/sdk-tracing.md

### Sampler

`Sampler` interface allows to create custom samplers which will return a
sampling `SamplingResult` based on information that is typically available just
before the `Span` was created.

#### ShouldSample

Returns the sampling Decision for a `Span` to be created.

> **Required arguments:**

Reviewer: Would it make sense to generalize this, in light of #31? In Jaeger we recently extended the API to allow sampling decisions to be made later in the span life cycle
```
  onCreateSpan(span: Span): SamplingDecision
  onSetOperationName(span: Span, operationName: string): SamplingDecision
  onSetTag(span: Span, key: string, value: any): SamplingDecision
```

Author: I think these are moments when the Sampler is called from the SDK perspective. I do agree that this is a nice to have config. Filed https://github.com/open-telemetry/opentelemetry-specification/issues/307

* `SpanContext` of a parent `Span`. Typically extracted from the wire. Can be
  `null`.
* `TraceId` of the `Span` to be created. It can be different from the `TraceId`
  in the `SpanContext`. Typically in situations when the `Span` to be created
  starts a new Trace.
* `SpanId` of the `Span` to be created.
* Name of the `Span` to be created.
* `SpanKind`
* Initial set of `Attributes` for the `Span` being constructed
* Collection of links that will be associated with the `Span` to be created.
Typically useful for batch operations, see
[Links Between Spans](overview.md#links-between-spans).

**Return value:**

It produces an output called `SamplingResult` which contains:

> * A `Decision` enum [`NOT_RECORD`, `RECORD`, `RECORD_AND_SAMPLED`].

Reviewer 2: please elaborate on what each of the enum value means

Author: done.

> * A `Decision` enum [`NOT_RECORD`, `RECORD`, `RECORD_AND_SAMPLED`].

Reviewer 2: Would it be better to rename `Decision` to  `SamplingDecision`?

Author: It depends on where do you have the class (happy to change if you want):
* As subclass of SamplingResult -> then call it just Decision
* As a standalone class probably -> SamplingDecision or SamplingResultDecision is a good name.

* A set of span Attributes that will also be added to the `Span`.

>   * These attributes will be added after the initial set of `Attributes`.

Reviewer 2: does "after" means that they will override the initial attributes or not? Unclear from the text

Author: If there is a key overlap and the implementation implements attributes as map yes.

Reviewer 2: it's unclear from the text as we never defined the order of attributes. I'd suggest to make text clearer

Author: Will just drop if there is no order, but I disagree because there is an explicit order even if it is not defined in the attributes section, also this document is not the right place to define that order.

### Sampler

`Sampler` interface allows to create custom samplers which will return a
sampling `SamplingResult` based on information that is typically available just
before the `Span` was created.

#### ShouldSample

Returns the sampling Decision for a `Span` to be created.

**Required arguments:**

* `SpanContext` of a parent `Span`. Typically extracted from the wire. Can be
  `null`.
* `TraceId` of the `Span` to be created. It can be different from the `TraceId`
  in the `SpanContext`. Typically in situations when the `Span` to be created
  starts a new Trace.
* `SpanId` of the `Span` to be created.
* Name of the `Span` to be created.
* `SpanKind`
* Initial set of `Attributes` for the `Span` being constructed
* Collection of links that will be associated with the `Span` to be created.
Typically useful for batch operations, see
[Links Between Spans](overview.md#links-between-spans).

**Return value:**

It produces an output called `SamplingResult` which contains:

> * A `Decision` enum:

Reviewer 2: ```suggestion
* A sampling `Decision`. One of the following enum values:
```

>   * `NOT_RECORD` - `IsRecording() == false` and all recorded trace events

Reviewer 2: ```suggestion
  * `NOT_RECORD` - `IsRecording() == false`, span will not be recorded and all events and attributes
```

  will be dropped.

>   * `RECORD` - `IsRecording() == true` AND `SampledFlag` is not set.

Reviewer 2: ```suggestion
  * `RECORD` - `IsRecording() == true`, but `Sampled` flag MUST NOT be set.
```

Reviewer 2: there is no `SampledFlag`, only flag called `sampled`

>   * `RECORD_AND_SAMPLED` - `IsRecording() == true` AND `SampledFlag` is set.

Reviewer 2: ```suggestion
  * `RECORD_AND_SAMPLED` - `IsRecording() == true` AND `Sampled` flag` MUST be set.
```

Description MUST NOT change over time and caller can cache the returned value.

### Built-in samplers

> These are the default samplers implemented in the OpenTelemetry SDK:

Reviewer 2: ```suggestion
These are the built-in samplers implemented in the OpenTelemetry SDK:
```

> * ALWAYS_ON

Reviewer 2: keep the `. Description MUST be `AlwaysSampleSampler`.` part at least

Author: done.

> * ALWAYS_ON

Reviewer 2: Specify that this sampler will be used as a default

Author: done.

* ALWAYS_OFF
* ALWAYS_PARENT

>   * Trust parent sampling decision (trusting and propagating parent `SampledFlag`).

Reviewer 2: the words propagating is confusing here. I'd just say - 

returns `RECORD_AND_SAMPLED` if `Sampled` flag is set to true on parent SpanContext and `NOT_RECORD` otherwise.

Author: done.

  * For root Spans (no parent available) returns `NOT_RECORD`.
* Probability

>   * Allows users to configure to ignore the parent `SampledFlag`.

Reviewer 2: I'd suggest to rephrase saying that the default behavior should be to trust the parent `Sampled` flag. However there should be configuration to opt out

Author: done.

>   * Allows users to configure if probability applies to "root spans",

Reviewer 2: I do not understand this. Please elaborate

Author: rephrased it.

  "root spans and remote parent", or "all spans".
    * Default is to apply only for "root spans and remote parent".

>   * Sample with 1/N probability

Reviewer 2: please suggest the algorithm. At least tell that this should be a function of a `TraceID`

Author: Add a section about implementation detail for this sampler, and left a TODO that mention the fact that is based on TraceId.

Description MUST NOT change over time and caller can cache the returned value.

### Built-in samplers

These are the default samplers implemented in the OpenTelemetry SDK:

> * ALWAYS_ON

Reviewer 3: IMO, this isn't a particularly clear name. Can we make it `ALWAYS_RETAIN` and the other be `ALWAYS_DROP` ?

Reviewer 3: I guess not?

Author: Please file an issue to not forget about this.

Reviewer 3: #314 done!
